src/java.base/share/classes/javax/crypto/KEMSpi.java

```diff
@@ -0,0 +1,245 @@
+
+/**
+ * This class defines the Service Provider Interface (SPI)
+ * for the {@link KEM} class. A security provider implements this interface
+ * to implement a KEM algorithm.
+ * <p>
+ * A KEM algorithm may contain a family of configurations.
+ * Different configurations accept different types of keys, uses different
+ * underlying cryptographic primitives, and supports different sizes of shared
+ * secrets and key encapsulation messages. A configuration must be decided
+ * by the KEM algorithm name, the key it uses, and an optional
+ * {@code AlgorithmParameterSpec} argument when creating an encapsulator or
+ * decapsulator. The result of either {@link #engineNewDecapsulator} or
+ * {@link #engineNewDecapsulator} must be of a single configuration,
+ * where its {@link EncapsulatorSpi#engineSecretSize()} and
+ * {@link EncapsulatorSpi#engineEncapsulationSize()} are constants.
+ * <p>
+ * A {@code KEMSpi} implementation should be immutable. It should be safe to
+ * call multiple {@code engineNewEncapsulator} and {@code engineNewDecapsulator}
+ * methods at the same time.
+ * <p>
+ * {@code EncapsulatorSpi} and {@code DecapsulatorSpi} implementations should
+ * be immutable. It should be safe to invoke multiple {@code encapsulate} and
+ * {@code decapsulate} methods at the same time. Each invocation of
+ * {@code encapsulate} should generate a new shared secret and key
+ * encapsulation message.
+ * <p>
+ * For example,
+ * {@snippet lang = java:
+ * public static class MyKEMImpl implements KEMSpi {
+ *
+ *     @Override
+ *     public KEMSpi.EncapsulatorSpi engineNewEncapsulator(PublicKey pk,
+ *             AlgorithmParameterSpec spec, SecureRandom secureRandom)
+ *             throws InvalidAlgorithmParameterException, InvalidKeyException {
+ *         if (!checkPublicKey(pk)) {
+ *             throw new InvalidKeyException("unsupported key");
+ *         }
+ *         if (!checkParameters(spec)) {
+ *             throw new InvalidAlgorithmParameterException("unsupported params");
+ *         }
+ *         return new MyEncapsulator(pk, spec, secureRandom);
+ *     }
+ *
+ *     class MyEncapsulator implements KEMSpi.EncapsulatorSpi {
+ *         MyEncapsulator(PublicKey pk, AlgorithmParameterSpec spec, SecureRandom secureRandom){
+ *             this.spec = spec != null ? spec : getDefaultParameters();
+ *             this.secureRandom = secureRandom != null ? secureRandom : getDefaultSecureRandom();
+ *             this.pk = pk;
+ *         }
+ *
+ *         @Override
+ *         public KEM.Encapsulated encapsulate(int from, int to, String algorithm) {
+ *             // ...
+ *         }
+ *
+ *         // ...
+ *     }
+ *
+ *     // ...
+ * }
+ * }
+ *
+ * @see KEM
+ * @since 21
+ */
+public interface KEMSpi {
+
+    /**
+     * The KEM encapsulator implementation, generated by
+     * {@link #engineNewEncapsulator}.
+     *
+     * @see KEM.Encapsulator
+     */
+    interface EncapsulatorSpi {
```

Comment:
Is it really necessary to define a EncapsulatorSpi?  It looks like an internal implementation for KEMSpi service, rather than a public individual interface that EncapsulatorSpi.getInstance() could be used.

Reply:
If I get it right, SPI means "service provider interface", which is normally for public service access by searching the registers.  In the provider implementation, a SPI implementation is normally registered so that it can be accessed.  I did not find the registration in the provider implementation.  I know it is right because it is not expected to be registered as it is not expected to be accessed directly.

> Security providers need to implement it and the KEM class uses it.
I know, but it is not necessarily designed as an SPI (it could be a normal interface that a provider need to implement and KEM class uses).  You can code like this, but I'm not sure if it is the common way.

Reply:
> So you think the name is misleading?

Yes.

> But then if you want to talk about one you have to say which class it is in, and this could be confusing and sometimes wrong. 

I did not get the idea yet to define it in KEPSpi.  It looks like that only one interface in the KEM should be sufficient.  The interface implementation details could be provider based.  It is OK to have an provider implementation implements an interface internally.

Reply:
> If the interface is only in `KEM`, then it needs a `provider()` method, but an implementation actually does not know what the provider is.

With "implementation", do you mean the javax/crypto/KEPSpi.java or src/java.base/share/classes/com/sun/crypto/provider/DHKEM.java?

If it is refer to KEPSpi.java, why KEPSpi.java need to know what the provider is?  Is it sufficient to use engineNewEncapsulator() to get the provider implementation?

If it is refer to DHKEM.java, I did not get the idea why the  provider is unknown.

> An implementation can be registered in any (or even multiple) providers.

I did not get the idea.  Why it is not registered in SunJCE?

I think you may have evaluated the following idea, but I'm not why it is not work.  I may missed something.  Would you mind explain in more details?

```
public final class KEM {
    interface Encapsulator {
        ...
        KEM.Encapsulated encapsulate(...);
        ...
    }

    public static KEM getInstance(String algorithm) {
        ...
    }

    // Search for the registered providers, return the 1st non-null provider.newEncapsulator() or throw exception.
    public Encapsulator newEncapsulator(PublicKey pk,
            AlgorithmParameterSpec spec, SecureRandom secureRandom)
        ...
    }
}

public interface KEMSpi {
    // A provider implementation will implement the KEM.Encapsulator 
    // interface internally.  If a provider does not support the parameters,
    // null or nil object will be returned.
    public KEM.Encapsulator newEncapsulator(PublicKey pk,
            AlgorithmParameterSpec spec, SecureRandom secureRandom);
}       

Use case:
    KEM.getInstance(DHKEM).newEncapsulator(...);
```

Reply:
> `DHKEM.java` is the implementation, and it does not know which provider it will be put into. It's inside the provider that calls `putService` or `put` to add an implementation there, not that the implementation registered itself in a provider.
> 
I did not get the idea.  Why DHKEM.java need register itself in a provider?  A DHKEM.java is a part of a provider, and the Provider implementation in the provider knows how to register DHKEM.

> If `getProvider()` is implemented inside the implementation, then it can only be attached to one provider. Also, do you expect it to return `new SunJCE()`? This means the `p` in `getInstance("DHKEM", p)` will be a different instance from the value returned by `getProvider()`. 

I did not get the idea.  How could it be possible to return different instances. `getInstance("DHKEM", p)` returns the DHKEM implementation in provider p.  The "DHKEM" string name here is not the DHKEM.java class in SunJCE provider.

Back to the question you have previously: 
> If the interface is only in KEM, then it needs a provider() method, but an implementation actually does not know what the provider is.

Why it is needed to know the provider of the interface?  Do you mean the Encapsulator provider could be different from the KEM provider?  That's, KEM provider is ProviderK, and the Encapsulator provider is ProviderE, and you want them work together?   Does it make sense that it is required that Encapsulator is an internal implementation of the KEM provider (i.e., both from the same provider)?

Reply:
For more details about option 2, please refer to https://github.com/openjdk/jdk/pull/13470/files.  The KEM.java and KEMSpi.java is pretty much the same except the clean up of En/Decapsulator(s) in this PR.

Reply:
> 2. Do validate parameters on your own (because no one else does)
> 
I would say: validate the parameters on your own, because not everyone else does.

Reply:
> I added another proposal in you PR at [#13470 (comment)](https://github.com/openjdk/jdk/pull/13470#issuecomment-1508915798).

I like the idea to use abstract class.  The concerns about provider() method get addressed.